### ISSUE TITLE
Drop long lived constant counters(e.g. zero rate ingress paths) at scrape time or by stream agg.

### ISSUE THREAD

**Issue by Sin4wd** [VictoriaMetrics/VictoriaMetrics#5288]
### Is your feature request related to a problem? Please describe

We are facing a problem with our scraped metrics from nginx and traefik ingress gateways. The number of time series exported is growing over time, unless we restart some pods. This is because some rarely used combination of labels are touched once in a while, yet multiple constant counter series remains until the next restart.
I understand that it's better for the application endpoints to let those long lived constant counters go away but It's nice to have the ability to limit these to prevent fluctuation in the storage.
I provide a simple example of such fluctuation on one of our traefik deployments:
```
sum by (job) (scrape_samples_scraped{job="traefik-metrics"})
```
![Screenshot 2023-11-05 at 1 29 45 PM](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/30809018/c68c0a51-a00a-4e85-8923-cf0dad79b982)
```
sum(rate(vm_rows_inserted_total{job=~"vmsingle-vmsingle"}[2m])) by (type) > 0
```
![Screenshot 2023-11-05 at 1 47 41 PM](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/30809018/26cc7274-9093-46f5-b515-920ef35dbc84)

The problem here is that usually these ingress metrics are the heaviest metrics of ours and handling such variation needs at least 2x-10x reserved ingestion capacity(cpu/memory) on the victoriametrics for some useless metrics.

### Describe the solution you'd like

There might be a way to drop some timeseries in the stream aggregation time based on a threshold.
for example:
- dropping those with 15m_increase <= 0
```
- match: 'traefik_router_requests_total'
  interval: 15m
  outputs: []
  drop:
    operator: le
    value: 0
```
- dropping those with 5m_increase <= 0 for 15m
```
- match: 'traefik_router_requests_total'
  interval: 5m
  outputs: []
  drop:
    operator: le
    value: 0
    for: 15m
```

### Describe alternatives you've considered

I have none.

**Comment by hagen1778** [comment]
> This is because some rarely used combination of labels are touched once in a while, yet multiple constant counter series remains until the next restart.

What are these metrics? Can it be solved on the exporter side?

**Comment by Sin4wd** [comment]
These are default exported metrics from traefik and nginx ingress controllers and there is no easy way to change their behavior.
For example:
```
traefik_router_requests_total
traefik_router_request_duration_seconds_bucket
nginx_ingress_controller_requests
nginx_ingress_controller_response_duration_seconds_bucket
```

**Comment by hagen1778** [comment]
@Sin4wd I think the problem needs to be fixed on the exporter side.
Adding a stream aggregation rule for this would use extra resources:
1. The amount of samples exposed by traefik and nginx controllers will remain on the same level, which means collector will need to fetch and parse all of them. Which is, according to your screenshots, is x2 increase over time.
2. To track whether something has changed or not on the time interval would require additional memory to be used by vmagent.

To sum up, adding a new aggregation rule just **moves** resource usage from vmstorage to vmagent. It doesn't **reduce** resource usage.
The proper solution would be to address cardinality issues on exporter side. Looking at the [traefik docs](https://github.com/traefik/traefik/blob/master/docs/content/observability/metrics/prometheus.md#prometheus) it seems possible to disable the verbosity of the metrics.

**Comment by Sin4wd** [comment]
@hagen1778 Thanks for your response. I totally understand your concern as I've mentioned:

> It's better for the application endpoints to let those long lived constant counters go away but It's nice to have the ability to limit these to prevent fluctuation in the storage.

I usually have something like `inactive_counter_lifetime=1h` in my code, however the problem exists in many well established softwares like traefik and nginx, in which case we are not in charge. As an observability solution like Victoriametrics, don't you think we should have a solution for such instance?

> It doesn't **reduce** resource usage.

Sorry, I didn't make my case clear enough. 
- It increases predictability of the vmstorage
- This approach drops garbage data right at the edge and doesn't bother middle way vmagents(thus reducing total load)
- It can increase query performace, reduce index size and reduce active timeseries 

I found some alternative solutions for the problems that might help others:
- In one case we ended up using logs for monitoring because logs where 100 lps compared to 300 k ingest rate and intermittent.
- In another case we have designed a monitoring pipeline with 2 million ingest rate capacity to support workload fluctuating between 500k to 1.6M
- I am thinking of writing a metric proxy that does the job and remotewrites to vmagent and would be grateful if you have a better idea.

Do you have any plans for vmagent to be a data transformer for metric timeseries? 
In this case, aside from this problem, the idea of filtering timeseries based on their values can be a great ability of the agent.
Feel free to close this issue if you do not have such plans.

**Comment by hagen1778** [comment]
> It increases predictability of the vmstorage

I'm not sure it does... From what I understand, these metrics can re-appear if someone hits specific path. So it really could be 100k samples/s now and 200k samples/s in one hour only because load pattern on nginx has changed. If it is acceptable for load pattern to change like this, then vmstorage should be ready to accept this data. 
Otherwise, [rate-limiting](https://docs.victoriametrics.com/vmagent.html#cardinality-limiter) should be set on vmagent side to limit the inbound only with level acceptable for vmstorage to handle.

> It can increase query performace, reduce index size and reduce active timeseries

It won't reduce index size, because once data was written at least one time - it remain in indexdb for the whole retention period.
It could lower cache sizes, though.

From what you're saying I see that problem exists. I'm just not sure that it should be solved on the receiver side. I need to think about this more.

**Comment by valyala** [comment]
Time series with constant values are compressed at much higher compression ratio (up by 1000x higher) than regular time series. This means such time series do not occupy too much additional disk space. Such time series also have much higher processing speed comparing to other time series during queries. Given these facts, it shouldn't be too expensive to store billions of constant samples for such time series.

As @hagen1778 mentioned above, when scrape target exposes unbound number of time series, then this may lead to [high cardinality](https://docs.victoriametrics.com/FAQ.html#what-is-high-cardinality) and [high churn rate](https://docs.victoriametrics.com/FAQ.html#what-is-high-churn-rate) issues. These issues cannot be avoided at [vmagent](https://docs.victoriametrics.com/vmagent.html) or VictoriaMetrics side even if dropping samples for constant counters, since VictoriaMetrics needs to register all the new [time series](https://docs.victoriametrics.com/keyConcepts.html#time-series) in the inverted index (aka `indexdb`). Basically it needs to store `(label="value") -> series_id` entries per each `label="value"` pair per each new time series in the `indexdb`. This increases `indexdb` size and may increase RAM usage alongside CPU usage at VictoriaMetrics side.

Given the facts above, there is little sense in dropping constant counters, since this will only reduce the network bandwidth usage between `vmagent` and `VictoriaMetrics`:

- It won't reduce the network bandwidth usage between scrape target and `vmagent`
- It won't reduce CPU and RAM usage at `vmagent` side, since it needs to read and parse all the metrics from the target before dropping constant counters. Moreover, it will increase RAM usage at `vmagent` side, since it needs to track all the ingested series in order to detect constant counters, which must be dropped.
- It won't reduce `indexdb` size, since every new constant counter should be registered at VictoriaMetrics side before being able to drop constant samples.
- It won't reduce RAM usage and CPU usage at VictoriaMetrics side related to handling of high cardinality and high churn rate.
